### PR TITLE
feat: Add support for async modules and async bootstrap

### PR DESCRIPTION
**Pull request by sjiang2019**
- Add `AsyncModule` which inherits from `Module`
- Add `AsyncBootstrapFewShot` which inherits from `BootstrapFewShot`

### CODE REVIEW

**Comment by okhat** [comment]
Hey, thanks a lot for opening this PR! We appreciate it. I left some comments.

Let me know your thoughts.

File: dspy/primitives/program.py [hunk 1]
```diff
@@ -23,41 +23,32 @@ def __init__(self):
...
 
-    #     for k, v in self.__dict__.items():
-    #         print(f"Copying attribute {k} of type {type(v)}...")
-    #         setattr(new_copy, k, copy.deepcopy(v, memo))
-    #         print("Done")
 
-    #     return new_copy
+class AsyncModule(Module):
+    async def __call__(self, *args, **kwargs):
+        return await self.forward(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (sjiang2019) [reply dspy/primitives/program.py:51]: Thanks for engaging! It depends on how you define `forward()`. The goal is just to allow you to do async ops, which could be LM-related or not (e.g. async db reads), in the module. For example,

```
class RAG(dspy.AsyncModule):
    def __init__(self):
        super().__init__()
    
    async def forward(self, question):
        search_entity_batches = await asyncio.gather(*[search_fn_1, search_fn_2, ...])
        # or could be async ranking, async ensemble answer generation, etc.
        ....
```

You're right in that it doesn't solve async for the current Predict and Retrieve abstractions, which use a sync LLM completion api call. Those would require a larger refactor to make async, but I think this is a reasonable escape hatch in the meantime.

**Comment by okhat** [comment]
This is so cool, thanks for the new updates! @lawliet19189 @detaos  is this something you'd be interested to test?

**Comment by okhat** [comment]
@sjiang2019 There's still a lot of content that may be redundant tho, I left some comments on those earlier, do you have thoughts on that

**Comment by sjiang2019** [comment]
@okhat Thanks! I'm slowly working on introducing an async path for LM calls as well. 

>There's still a lot of content that may be redundant tho, I left some comments on those earlier, do you have thoughts on that

I can't seem to see your earlier comments for some reason (I only see the one about async), but will take a stab at reducing the redundancies.

**Comment by Vasilije1990** [comment]
@okhat is there some progress here? We need this feature and are thinking of redoing it ourselves and opening a new PR without changing the code that is not relevant. wdyt?

**Comment by arnavsinghvi11** [comment]
@Vasilije1990 Feel free to go ahead and create a new PR for async support! (Don't think any progress has been made to this PR but feel free to push onto it or create a new one, whichever works best)

**Comment by vaibhavnayel** [comment]
Hey @arnavsinghvi11,  is async support part of the development roadmap? I don't see any active or closed PRs about this. I would be happy to pitch in if there is interest.

**Comment by owen-deepskill** [comment]
@okhat I wonder if there is a plan for DSPy to natively support AsyncOpenAI and AsyncAnthropic? I know we can use sync_to_async, but I guess it's not exactly same with the native async function. Thanks!

**Comment by CyrusNuevoDia** [comment]
@Vasilije1990 @vaibhavnayel @NumberChiffre @owen-deepskill See #1806 for `dspy.asyncify` support!

**Comment by owen-deepskill** [comment]
@CyrusNuevoDia Awesome! Many thanks for the update!

**Pull request closed without merging**